xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -377,14 +387,24 @@ void watchEndpointData(String clusterName, EndpointWatcher watcher) {
       logger.log(
           XdsLogLevel.DEBUG,
           "Endpoint resource for cluster {0} is known to be absent.", clusterName);
-      watcher.onResourceDoesNotExist(clusterName);
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          watcher.onResourceDoesNotExist(clusterName);
+        }
+      });
       return;
     }
     if (clusterNamesToEndpointUpdates.containsKey(clusterName)) {
       logger.log(
           XdsLogLevel.DEBUG,
           "Retrieve endpoints info for cluster {0} from local cache.", clusterName);
-      watcher.onEndpointChanged(clusterNamesToEndpointUpdates.get(clusterName));
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          watcher.onEndpointChanged(clusterNamesToEndpointUpdates.get(clusterName));
```

Comment:
Do you need to check `clusterNamesToEndpointUpdates.containsKey(clusterName` again?